crates/cnidarium-component/src/action_handler.rs

```diff
@@ -44,33 +44,45 @@ pub trait ActionHandler {
     /// be run in parallel across all transactions in a block.
     async fn check_stateless(&self, context: Self::CheckStatelessContext) -> Result<()>;
 
-    /// Performs all of this action's stateful validity checks.
+    /// Performs those stateful validity checks that can be performed against a
+    /// historical state snapshot.
     ///
     /// This method provides read access to a snapshot of the `State` prior to
-    /// transaction execution.  Because it does not give write access, it's safe
-    /// to run the `ActionHandler` implementations for all actions in a given
-    /// transaction in parallel.
+    /// transaction execution.  It is intended to be run in parallel across all
+    /// actions within a transaction.
     ///
-    /// As much work as possible should be done in `check_stateful`, as it can
-    /// be run in parallel across all actions within a transaction.
+    /// # Warning
+    ///
+    /// Misuse of this method creates TOCTOU vulnerabilities. Checks performed
+    /// in this method must be valid if they are performed against a _prior_
+    /// state, as another action in the same transaction may execute first and
+    /// change the state.
+    ///
+    /// Checks performed in this phase should have a justification for why they
+    /// are safe to run in parallel with other actions in the same transaction,
+    /// and the default behavior should be to perform checks in
+    /// [`ActionHandler::check_and_execute`].
```

Comment:
appreciate the additional information added in the comments here :heavy_check_mark: